functions/access.js

```diff
@@ -0,0 +1,136 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require("./auth");
+const { uuid } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  if (!venueId || !uid || !token) return false;
+
+  const venueRef = admin.firestore().collection("venues").doc(venueId);
+  const accessRef = admin.firestore().collection("accessgranted").doc(uid);
+
+  return await admin
+    .firestore()
+    .runTransaction(async (transaction) => {
+      const venue = await transaction.get(venueRef);
+      const granted = await transaction.get(accessRef);
+
+      if (!venue.exists) {
+        return false;
+      }
+      if (!granted.exists || !granted.tokens) {
+        return false;
+      }
+
+      if (Object.keys(granted.tokens).includes(token)) {
+        // Record that the token was checked
+        if (!granted.tokens[token].usedAt) {
+          granted.tokens[token].usedAt = [];
+        }
+        granted.tokens[token].usedAt.push(Date.now());
+        transaction.update(accessRef, granted);
+
+        return true;
+      }
+
+      return false;
+    })
+    .catch((e) => {
+      console.log("Transaction failure:", e);
```

Comment:
`console.error`?

[OOS] It would also be good to add `Bugsnag` here at some stage.. but I don't think we have it setup for the functions yet, so that's probably OOS of this PR.